From e6caa0f7991b00bb3a61dccf0b1022ee857d96c7 Mon Sep 17 00:00:00 2001
From: Samuel Pitoiset <samuel.pitoiset@gmail.com>
Date: Fri, 6 Dec 2024 16:25:18 +0100
Subject: [PATCH 4/6] radv: add radv_disable_dcc_stores and enable for Indiana
 Jones: The Great Circle

Likely a game bug but can't be 100% sure because the game uses RT by
default and renderdoc still doesn't have support for it.

Cc: mesa-stable
Signed-off-by: Samuel Pitoiset <samuel.pitoiset@gmail.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/32528>
(cherry picked from commit e3d1f27b31816ce9e8fddfe9c669059893cd8934)

Conflicts:
	src/util/00-radv-defaults.conf
---
 .gitignore                     | 1 +
 src/amd/vulkan/radv_image.c    | 4 ++++
 src/amd/vulkan/radv_instance.c | 2 ++
 src/amd/vulkan/radv_instance.h | 1 +
 src/util/00-radv-defaults.conf | 6 ++++++
 src/util/driconf.h             | 4 ++++
 6 files changed, 18 insertions(+)

diff --git a/.gitignore b/.gitignore
index 30f4e69ea7d..3f389086fbf 100644
--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@
 *.out
 /build
 .venv/
+.pick_status.json
diff --git a/src/amd/vulkan/radv_image.c b/src/amd/vulkan/radv_image.c
index 5cc04237679..354e7951b31 100644
--- a/src/amd/vulkan/radv_image.c
+++ b/src/amd/vulkan/radv_image.c
@@ -295,6 +295,10 @@ radv_use_dcc_for_image_early(struct radv_device *device, struct radv_image *imag
    if (instance->drirc.disable_dcc_mips && pCreateInfo->mipLevels > 1)
       return false;
 
+   /* Force disable DCC for stores to workaround game bugs. */
+   if (instance->drirc.disable_dcc_stores && (pCreateInfo->usage & VK_IMAGE_USAGE_STORAGE_BIT))
+      return false;
+
    /* DCC MSAA can't work on GFX10.3 and earlier without FMASK. */
    if (pCreateInfo->samples > 1 && pdev->info.gfx_level < GFX11 && (instance->debug_flags & RADV_DEBUG_NO_FMASK))
       return false;
diff --git a/src/amd/vulkan/radv_instance.c b/src/amd/vulkan/radv_instance.c
index df2ae2cd307..f81edfd8663 100644
--- a/src/amd/vulkan/radv_instance.c
+++ b/src/amd/vulkan/radv_instance.c
@@ -153,6 +153,7 @@ static const driOptionDescription radv_dri_options[] = {
       DRI_CONF_RADV_DISABLE_TC_COMPAT_HTILE_GENERAL(false)
       DRI_CONF_RADV_DISABLE_DCC(false)
       DRI_CONF_RADV_DISABLE_DCC_MIPS(false)
+      DRI_CONF_RADV_DISABLE_DCC_STORES(false)
       DRI_CONF_RADV_DISABLE_ANISO_SINGLE_LEVEL(false)
       DRI_CONF_RADV_DISABLE_TRUNC_COORD(false)
       DRI_CONF_RADV_DISABLE_SINKING_LOAD_INPUT_FS(false)
@@ -265,6 +266,7 @@ radv_init_dri_options(struct radv_instance *instance)
    instance->drirc.vk_require_astc = driQueryOptionb(&instance->drirc.options, "vk_require_astc");
 
    instance->drirc.disable_dcc_mips = driQueryOptionb(&instance->drirc.options, "radv_disable_dcc_mips");
+   instance->drirc.disable_dcc_stores = driQueryOptionb(&instance->drirc.options, "radv_disable_dcc_stores");
 }
 
 static const struct vk_instance_extension_table radv_instance_extensions_supported = {
diff --git a/src/amd/vulkan/radv_instance.h b/src/amd/vulkan/radv_instance.h
index 876461b642b..57d6bdc5a48 100644
--- a/src/amd/vulkan/radv_instance.h
+++ b/src/amd/vulkan/radv_instance.h
@@ -72,6 +72,7 @@ struct radv_instance {
       bool vk_require_etc2;
       bool vk_require_astc;
       bool disable_dcc_mips;
+      bool disable_dcc_stores;
       char *app_layer;
       uint8_t override_graphics_shader_version;
       uint8_t override_compute_shader_version;
diff --git a/src/util/00-radv-defaults.conf b/src/util/00-radv-defaults.conf
index 5bf2bedc960..d125e24ede8 100644
--- a/src/util/00-radv-defaults.conf
+++ b/src/util/00-radv-defaults.conf
@@ -109,6 +109,12 @@ Application bugs worked around in this file:
             <option name="radv_legacy_sparse_binding" value="true" />
         </application>
 
+        <application name="Indiana Jones: The Great Circle" application_name_match="TheGreatCircle">
+            <option name="radv_zero_vram" value="true" />
+            <option name="radv_legacy_sparse_binding" value="true" />
+            <option name="radv_disable_dcc_stores" value="true" />
+        </application>
+
         <application name="DOOM (2016)" application_name_match="DOOM$">
             <option name="radv_disable_dcc" value="true" />
         </application>
diff --git a/src/util/driconf.h b/src/util/driconf.h
index 769d2dfd7d5..f988ad7fc8c 100644
--- a/src/util/driconf.h
+++ b/src/util/driconf.h
@@ -696,6 +696,10 @@
    DRI_CONF_OPT_B(radv_disable_dcc_mips, def, \
                   "Disable DCC for color images with mips")
 
+#define DRI_CONF_RADV_DISABLE_DCC_STORES(def) \
+   DRI_CONF_OPT_B(radv_disable_dcc_stores, def, \
+                  "Disable DCC for color storage images")
+
 #define DRI_CONF_RADV_DISABLE_ANISO_SINGLE_LEVEL(def) \
   DRI_CONF_OPT_B(radv_disable_aniso_single_level, def, \
                  "Disable anisotropic filtering for single level images")
-- 
2.47.1

